crates/cext/src/transpiler/passes/sabre_layout.rs

```diff
@@ -0,0 +1,132 @@
+/// of times until it finds an initial_layout that reduces full routing cost.
+///
+/// This method exploits the reversibility of quantum circuits, and tries to include global circuit information in the
+/// choice of initial_layout.
+///
+/// This pass will run both layout and routing and will transform the circuit so that the layout is applied to the input
+/// (meaning that the output circuit will have ancilla qubits allocated for unused qubits on the coupling map and the
+/// qubits will be reordered to match the mapped physical qubits) and then routing will be applied. This is done because
+/// the pass will run parallel seed trials with different random seeds for selecting the random initial layout and then
+/// selecting the routed output which results in the least number of swap gates needed. This final
+/// swap calculation is the same as performing a final routing, so it's more efficient to apply it
+/// after computing it.
+///
+/// This function is multithreaded and will launch a thread pool with threads equal to the number
+/// of CPUs by default. You can tune the number of threads with the ``RAYON_NUM_THREADS``
+/// environment variable. For example, setting ``RAYON_NUM_THREADS=4`` would limit the thread pool
+/// to 4 threads.
+///
+/// # References
+///
+/// [1] Henry Zou and Matthew Treinish and Kevin Hartman and Alexander Ivrii and Jake Lishman.
+/// "LightSABRE: A Lightweight and Enhanced SABRE Algorithm"
+/// [arXiv:2409.08368](https://doi.org/10.48550/arXiv.2409.08368)
+///
+/// [2] Li, Gushu, Yufei Ding, and Yuan Xie. "Tackling the qubit mapping problem
+/// for NISQ-era quantum devices." ASPLOS 2019.
+/// [`arXiv:1809.02573](https://arxiv.org/pdf/1809.02573.pdf)
+///
+/// @param circuit A pointer to the circuit to run SabreLayout on
+/// @param target A pointer to the target to run SabreLayout on
+/// @param max_iterations The number of forward-backward iterations in the
+///     sabre routing algorithm
+/// @param num_swap_trials The number of trials to run of the sabre routing
+///     algorithm for each iteration. When > 1 the trial that routing trial
+///     that results in the output with the fewest swap gates will be selected.
+/// @param num_random_trials The number of random layout trials to run. The trial
+///     that results in the output with the fewest swap gates will be selected.
+/// @param seed A seed value for the pRNG used internally. If the value is negative
+///     the RNG will be seeded from system entropy.
+///
+/// @return The transpile layout that describes the layout and output permutation caused
+///     by the pass
+///
+/// # Safety
+///
+/// Behavior is undefined if ``circuit`` or ``target`` is not a valid, non-null pointer to a ``QkCircuit`` and ``QkTarget``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_transpiler_pass_standalone_sabre_layout(
+    circuit: *mut CircuitData,
```

Comment:
Do we have any C-level objects that borrow a reference from a pointer?  If so, we probably should get in the habit of documenting in passes like this one that calling frees the underlying `CircuitData` object, and so can only be called with the last extant reference to a `CircuitData`.

Reply:
Added in: https://github.com/Qiskit/qiskit/pull/14711/commits/6209b8d2ebdccea02806741a2cb0fed6cecd5e18